src/irmin/commit.ml

```diff
@@ -25,43 +25,75 @@ module Log = (val Logs.src_log src : Logs.LOG)
 module Maker (Info : Info.S) = struct
   module Info = Info
 
-  module Make (H : Type.S) = struct
+  module Make
+      (H : Type.S)
+      (N : Key.S with type hash = H.t)
+      (C : Key.S with type hash = H.t) =
+  struct
     module Info = Info
 
-    type hash = H.t [@@deriving irmin ~compare]
+    type node_key = N.t [@@deriving irmin ~compare]
+    type commit_key = C.t [@@deriving irmin]
 
-    type t = { node : hash; parents : hash list; info : Info.t }
+    type t = { node : node_key; parents : commit_key list; info : Info.t }
     [@@deriving irmin]
 
     let parents t = t.parents
     let node t = t.node
     let info t = t.info
+    let compare_hash = Type.(unstage (compare H.t))
+    let compare_commit x y = compare_hash (C.to_hash x) (C.to_hash y)
 
     let v ~info ~node ~parents =
-      let parents = List.fast_sort compare_hash parents in
+      let parents = List.fast_sort compare_commit parents in
       { node; parents; info }
+
+    module Pre_hash = struct
+      type t = { p_node : H.t; p_parents : H.t list; p_info : Info.t }
+      [@@deriving irmin]
+
+      let v t =
+        {
+          p_node = N.to_hash t.node;
+          p_parents = List.map C.to_hash t.parents;
```

Comment:
It would be great if we could call the pre_hash iterator without having to build that intermediate `map`? Not sure what this API would look like though... maybe passing an `pre_hash_of_hash` function as a callback. Here it's not a big deal as the number of parents is always low, but for nodes it might have an impact.

Reply:
Looking at this again, I think this `Pre_hash` module indirection is now unnecessary thanks to https://github.com/mirage/repr/pull/71. The backend must just ensure that keys are pre-hashed as hashes – I'll add that to the docs somewhere & strip this out.

Reply:
Done.